test/renderer/stores/index.spec.js

```diff
@@ -1772,6 +1787,177 @@ describe('Store', () => {
+      Store.state.connections = { ...connections };
+      Store.state.isConnecting = false;
+      Store.state.isConnected = false;
+      Store.state.currentConnection = connection;
+
+      connectionAttemptId = uuidv4();
+      Store.connectionAttemptId = connectionAttemptId;
+      Store.dataService = null;
+      Store.connectingDataService = null;
+
+      appRegistryEmitStub = sinon.fake();
+      sinon.replace(
+        Store,
+        'appRegistry',
+        {
+          emit: appRegistryEmitStub
+        }
+      );
+    });
+
+    afterEach(async() => {
+      if (Store.dataService) {
+        try {
+          await Store.dataService.disconnect();
+        } catch (err) { /* */ }
+        Store.dataService = null;
+      }
+
+      if (Store.connectingDataService) {
+        try {
+          await Store.connectingDataService.disconnect();
+        } catch (err) { /* */ }
+        Store.connectingDataService = null;
+      }
+
+      sinon.restore();
+    });
+
+    it('connects to the database and sets the dataService on the store', async() => {
+      await Store._connect(connection, connectionAttemptId);
+
+      expect(Store.dataService).to.not.equal(null);
+      expect(Store.state.isConnected).to.equal(true);
+      expect(appRegistryEmitStub.calledOnce).to.equal(true);
+    });
+
+    it('shows the progress bar when it successfully connects', async() => {
+      const spyShow = sinon.spy(
+        Store.StatusActions,
+        'showIndeterminateProgressBar'
+      );
+
+      await Store._connect(connection, connectionAttemptId);
+
+      expect(spyShow.calledOnce).to.equal(true);
+    });
+
+    it('does not show the progress bar when it errors when connecting', async() => {
+      const spyShow = sinon.spy(
+        Store.StatusActions,
+        'showIndeterminateProgressBar'
+      );
+
+      let finishedConnecting = false;
+      const startConnecting = async() => {
+        await Store._connect({
+          port: 29799 // Hopefully not in use.
+        }, connectionAttemptId);
+
+        finishedConnecting = true;
+      };
+
+      startConnecting();
+
+      await ensureResult(
+        3,
+        () => Store.connectingDataService,
+        () => Store.connectingDataService !== null,
+        'Never started connecting to failing connection.'
+      );
+
+      Store.connectingDataService.disconnect();
+
+      await ensureResult(
+        3,
+        () => finishedConnecting,
+        () => finishedConnecting,
+        'Never finished connecting to failing connection.'
+      );
+
+      expect(Store.state.isConnected).to.equal(false);
+      expect(Store.connectingDataService).to.equal(null);
+      expect(Store.dataService).to.equal(null);
+      expect(spyShow.calledOnce).to.equal(false);
+    });
+
+    it('does not attempt to connect when the connectionAttemptId does not match the one on the store', async() => {
+      const differentConnectionAttemptId = uuidv4();
```

Comment:
Isn't there a way to just have only one connection in progress at the same time?